### ISSUE TITLE
Long running Genrule can't finish when "jobs" bigger then 1, thus compilation errors out.

### ISSUE THREAD

**Issue by agostini01** [bazelbuild/bazel#10384]
### [BUG?] Description of the problem:

**TLDR**: Long running genrule dependency is not allowed to finish before the compilation of "who depends on it" starts. Thus compilation errors out as the generated files have not been created yet. Genrule takes around 20s with 4 cores to complete.

Maybe not standard procedure but I am trying to compile a non bazel library project once as a dependency of one of my binaries covered under bazel.

For that, 

0. Downloaded my project with tf_http_archive
1. I used a genrule that makes a call to cmake and make. 
2. Then created a cc_library depend on this genrule output file
3. Then made the binary depend on this library.

If I execute build command with multiple jobs, genrule does not have time to finish, because the compilation of the binary starts straight away.

Scheduling bazel build with one job or having a long queue of actions that delay the build of my binary allows the genrule to finish and the compilation to successfully complete.

This is my first week with bazel and I am not sure how to debug or what is going wrong.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

See files below.

### What operating system are you running Bazel on?

Docker tensorflow/tensorflow:nightly-custom-op-ubuntu16 as of Dec/03/2019. bazel 1.1.0

### What's the output of `bazel info release`?

```
INFO: Options provided by the client:
  Inherited 'common' options: --isatty=1 --terminal_columns=275
INFO: Reading rc options for 'info' from /working_dir/tensorflow/.bazelrc:
  Inherited 'build' options: --apple_platform_type=macos --define framework_shared_object=true --define open_source_build=true --java_toolchain=//third_party/toolchains/java:tf_java_toolchain --host_java_toolchain=//third_party/toolchains/java:tf_java_toolchain --define=use_fast_cpp_protos=true --define=allow_oversize_protos=true --spawn_strategy=standalone -c opt --announce_rc --define=grpc_no_ares=true --noincompatible_remove_legacy_whole_archive --enable_platform_specific_config --config=v2
INFO: Found applicable config definition build:v2 in file /working_dir/tensorflow/.bazelrc: --define=tf_api_version=2 --action_env=TF2_BEHAVIOR=1
INFO: Found applicable config definition build:linux in file /working_dir/tensorflow/.bazelrc: --copt=-w --define=PREFIX=/usr --define=LIBDIR=$(PREFIX)/lib --define=INCLUDEDIR=$(PREFIX)/include --cxxopt=-std=c++14 --host_cxxopt=-std=c++14
release 1.1.0
```

###  Have you found anything relevant by searching the web?

Tried the `https://github.com/bazelbuild/rules_foreign_cc` but it does not fit my needs as it is complex to integrate in the framework (not using http_archive but tf_http_archive).

### Files involved:

workspace.bzl
```
    tf_http_archive(                                                               
        name = "systemc",                                                          
        build_file = clean_dep("//third_party:systemc.BUILD"),                     
        sha256 = "5781b9a351e5afedabc37d145e5f7edec08f3fd5de00ffeb8fa1f3086b1f7b3f",
        urls = [                                                                   
            "https://www.accellera.org/images/downloads/standards/systemc/systemc-2.3.3.tar.gz",
            "https://www.accellera.org/images/downloads/standards/systemc/systemc-2.3.3.tar.gz",
        ],                                                                         
    )
```

BUILD
```
cc_binary(                                                                         
    name = "systemc_model",                                                        
    srcs = [                                                                       
        "systemc_main.sc.cc",                                                      
    ],                                                                                                                                        
    deps = [                                                                                                                                  
        "@systemc//:systemc",                                                      
    ],                                                                             
)  
```

third_party/systemc.BUILD
```
licenses(["notice"])                                                               
                                                                                   
package(default_visibility = ["//visibility:public"])                              
                                                                                   
genrule(                                                                           
    name = "libsystemc",                                                           
    srcs = [],                                                                     
    outs = ["systemc-2.3.3/install/lib/libsystemc.a"],                             
    cmd =                                                                          
        "cmake -DCMAKE_INSTALL_PREFIX=external/systemc/systemc-2.3.3/install -DCMAKE_CXX_STANDARD=14 -DCMAKE_INSTALL_INCLUDEDIR=include/systemc -DBUILD_SHARED_LIBS=off -Bexternal/systemc/systemc-2.3.3/build -Hexternal/systemc/systemc-2.3.3 &&" +
        "make -C external/systemc/systemc-2.3.3/build install -j4 &&" +            
        "cp external/systemc/systemc-2.3.3/install/lib/libsystemc.a $@",           
)                                                                                  
                                                                                   
cc_library(                                                                        
    name = "systemc",                                                              
    srcs = ["systemc-2.3.3/install/lib/libsystemc.a"],                             
    hdrs = glob([                                                                  
        "systemc-2.3.3/install/include/system.h",                                  
    ]),                                                                            
    copts = ["std=c++14"],                                                         
    data = [":libsystemc"],                                                        
    includes = [                                                                   
        "systemc-2.3.3/install/include",                                           
        "systemc-2.3.3/install/include/systemc",                                   
    ],                                                                             
) 
```

Success messages when running `bazel build --jobs 1 --explain=file.txt --verbose_explanations tensorflow/lite/tools/systemc:systemc_model`

```
Build options: --apple_platform_type=macos --define='framework_shared_object=true' --define='open_source_build=true' --java_toolchain=//third_party/toolchains/java:tf_java_toolchain --host_java_toolchain=//third_party/toolchains/java:tf_java_toolchain --define='use_fast_cpp_protos=true' --define='allow_oversize_protos=true' --spawn_strategy=standalone --compilation_mode=opt --announce_rc --define='grpc_no_ares=true' --noincompatible_remove_legacy_whole_archive --enable_platform_specific_config --copt=-w --define='PREFIX=/usr' --define='LIBDIR=$(PREFIX)/lib' --define='INCLUDEDIR=$(PREFIX)/include' --cxxopt='-std=c++14' --host_cxxopt='-std=c++14' --config=v2 --define='tf_api_version=2' --action_env='TF2_BEHAVIOR=1' --jobs=1 --explain=file.txt --verbose_explanations
Executing action 'BazelWorkspaceStatusAction stable-status.txt': unconditional execution is requested.
Executing action 'Executing genrule @systemc//:libsystemc': no entry in the cache (action is new).
Executing action 'Creating source manifest for //tensorflow/lite/tools/systemc:systemc_model': no entry in the cache (action is new).
Executing action 'Creating runfiles tree bazel-out/k8-opt/bin/tensorflow/lite/tools/systemc/systemc_model.runfiles': no entry in the cache (action is new).
Executing action 'Writing file tensorflow/lite/tools/systemc/systemc_model-2.params': no entry in the cache (action is new).
Executing action 'Compiling tensorflow/lite/tools/systemc/systemc_main.sc.cc': no entry in the cache (action is new).
Executing action 'Linking tensorflow/lite/tools/systemc/systemc_model': no entry in the cache (action is new).
```

Error messages when running unlimited jobs `bazel build --explain=file.txt --verbose_explanations tensorflow/lite/tools/systemc:systemc_model`

```
Build options: --apple_platform_type=macos --define='framework_shared_object=true' --define='open_source_build=true' --java_toolchain=//third_party/toolchains/java:tf_java_toolchain --host_java_toolchain=//third_party/toolchains/java:tf_java_toolchain --define='use_fast_cpp_protos=true' --define='allow_oversize_protos=true' --spawn_strategy=standalone --compilation_mode=opt --announce_rc --define='grpc_no_ares=true' --noincompatible_remove_legacy_whole_archive --enable_platform_specific_config --copt=-w --define='PREFIX=/usr' --define='LIBDIR=$(PREFIX)/lib' --define='INCLUDEDIR=$(PREFIX)/include' --cxxopt='-std=c++14' --host_cxxopt='-std=c++14' --config=v2 --define='tf_api_version=2' --action_env='TF2_BEHAVIOR=1' --explain=file.txt --verbose_explanations
Executing action 'BazelWorkspaceStatusAction stable-status.txt': unconditional execution is requested.
Executing action 'Creating source manifest for //tensorflow/lite/tools/systemc:systemc_model': no entry in the cache (action is new).
Executing action 'Writing file tensorflow/lite/tools/systemc/systemc_model-2.params': no entry in the cache (action is new).
Executing action 'Executing genrule @systemc//:libsystemc': no entry in the cache (action is new).
Executing action 'Compiling tensorflow/lite/tools/systemc/systemc_main.sc.cc': no entry in the cache (action is new).
Executing action 'Creating runfiles tree bazel-out/k8-opt/bin/tensorflow/lite/tools/systemc/systemc_model.runfiles': no entry in the cache (action is new).
```

and 
```
Starting local Bazel server and connecting to it...
INFO: Writing tracer profile to '/home/developer/.cache/bazel/_bazel_developer/881ae704c1434b3a766ac83e64e752e0/command.profile.gz'
INFO: Options provided by the client:
  Inherited 'common' options: --isatty=1 --terminal_columns=275
INFO: Reading rc options for 'build' from /working_dir/tensorflow/.bazelrc:
  'build' options: --apple_platform_type=macos --define framework_shared_object=true --define open_source_build=true --java_toolchain=//third_party/toolchains/java:tf_java_toolchain --host_java_toolchain=//third_party/toolchains/java:tf_java_toolchain --define=use_fast_cpp_protos=true --define=allow_oversize_protos=true --spawn_strategy=standalone -c opt --announce_rc --define=grpc_no_ares=true --noincompatible_remove_legacy_whole_archive --enable_platform_specific_config --config=v2
INFO: Found applicable config definition build:v2 in file /working_dir/tensorflow/.bazelrc: --define=tf_api_version=2 --action_env=TF2_BEHAVIOR=1
INFO: Found applicable config definition build:linux in file /working_dir/tensorflow/.bazelrc: --copt=-w --define=PREFIX=/usr --define=LIBDIR=$(PREFIX)/lib --define=INCLUDEDIR=$(PREFIX)/include --cxxopt=-std=c++14 --host_cxxopt=-std=c++14
DEBUG: Rule 'io_bazel_rules_docker' indicated that a canonical reproducible form can be obtained by modifying arguments shallow_since = "1556410077 -0400"
DEBUG: Call stack for the definition of repository 'io_bazel_rules_docker' which is a git_repository (rule definition at /home/developer/.cache/bazel/_bazel_developer/881ae704c1434b3a766ac83e64e752e0/external/bazel_tools/tools/build_defs/repo/git.bzl:195:18):
 - /home/developer/.cache/bazel/_bazel_developer/881ae704c1434b3a766ac83e64e752e0/external/bazel_toolchains/repositories/repositories.bzl:37:9
 - /working_dir/tensorflow/WORKSPACE:37:1
INFO: Analyzed target //tensorflow/lite/tools/systemc:systemc_model (19 packages loaded, 69 targets configured).
INFO: Found 1 target...
INFO: Writing explanation of rebuilds to 'file.txt'
ERROR: /working_dir/tensorflow/tensorflow/lite/tools/systemc/BUILD:22:1: C++ compilation of rule '//tensorflow/lite/tools/systemc:systemc_model' failed (Exit 1)
tensorflow/lite/tools/systemc/systemc_main.sc.cc:3:29: fatal error: systemc/systemc.h: No such file or directory
compilation terminated.
Target //tensorflow/lite/tools/systemc:systemc_model failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 3.761s, Critical Path: 0.04s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
```

**Comment by agostini01** [comment]
Note, I can run:
```
bazel build tensorflow/lite/tools/systemc:systemc_model
```
Wait for the error.
Then run: 
```
bazel build -jobs 1 tensorflow/lite/tools/systemc:systemc_model
```
And compilation completes.

**Comment by iirina** [comment]
Are all the outputs declared? For example is `systemc/systemc.h` declared as an output so Bazel knows to wait for it to be built?

**Comment by agostini01** [comment]
Sorry for the belated answer.

It was not. Tomorrow morning I will do the test declaring the header.
At that point I did not understand Bazel that well.

Currently, this is what is on my "third party" tools bazel build file.

```
package(default_visibility = ["//visibility:public"])                              
                                                                                   
genrule(                                                                           
    name = "libsystemc",                                                           
    srcs = [],                                                                     
    outs = ["systemc-2.3.3/install/lib/libsystemc.a"],                             
    cmd =                                                                          
        "cmake -DCMAKE_INSTALL_PREFIX=external/systemc/systemc-2.3.3/install -DCMAKE_CXX_STANDARD=14 -DCMAKE_INSTALL_INCLUDEDIR=include/       …systemc -DBUILD_SHARED_LIBS=off -Bexternal/systemc/systemc-2.3.3/build -Hexternal/systemc/systemc-2.3.3 &&" +
        "make -C external/systemc/systemc-2.3.3/build install -j4 &&" +            
        "cp external/systemc/systemc-2.3.3/install/lib/libsystemc.a $@",           
)                                                                                  
                                                                                   
cc_library(                                                                        
    name = "systemc",                                                              
    srcs = ["systemc-2.3.3/install/lib/libsystemc.a"],                             
    hdrs = glob([                                                                  
        "systemc-2.3.3/install/include/system.h",                                  
    ]),                                                                            
    copts = ["std=c++14"],                                                         
    data = [":libsystemc"],                                                        
    includes = [                                                                   
        "systemc-2.3.3/install/include",                                           
        "systemc-2.3.3/install/include/systemc",                                   
    ],                                                                             
) 
```

**Comment by agostini01** [comment]
Simply adding the list of headers to `outs` makes the compilation error out with:

```
ERROR: /home/developer/.cache/bazel/_bazel_developer/881ae704c1434b3a766ac83e64e752e0/external/systemc/BUILD.bazel:38:1: in cmd attribute of genrule rule @systemc//:libsystemc: variable '$@' : more than one output file
ERROR: Analysis of target '//tensorflow/lite/examples/systemc:sc_example_test' failed; build aborted: Analysis of target '@systemc//:libsystemc' failed; build aborted
```

Which indicates that I cannot use the `$@` symbol anymore. 
This symbol previously translated to: 

```
bazel-out/k8-opt/bin/external/systemc/systemc-2.3.3/install/lib/libsystemc.a
```

I am not sure if this pre-appended path is always the same, or will change as I build more versions of the library.

When hardcoding the `@$` symbol and adding some of the (required) generated headers

```
genrule(                                                                           
    name = "libsystemc",                                                           
    srcs = [],                                                                     
    outs = ["systemc-2.3.3/install/lib/libsystemc.a",                              
            "systemc-2.3.3/install/include/systemc/systemc.h",                     
            ],                                                                     
    cmd =                                                                          
        "cmake -DCMAKE_INSTALL_PREFIX=external/systemc/systemc-2.3.3/install -DCMAKE_CXX_STANDARD=14 -DCMAKE_INSTALL_INCLUDEDIR=include/systemc -DBUILD_SHARED_LIBS=off -Bexternal/systemc/systemc-2.3.3/build -Hexternal/systemc/systemc-2.3.3 &&" +
        "make -C external/systemc/systemc-2.3.3/build install -j4 &&" +            
        "cp external/systemc/systemc-2.3.3/install/lib/libsystemc.a bazel-out/k8-opt/bin/external/systemc/systemc-2.3.3/install/lib/libsystemc.a &&" +
        "cp external/systemc/systemc-2.3.3/install/include/systemc/systemc.h bazel-out/k8-opt/bin/external/systemc/systemc-2.3.3/install/include/systemc/systemc.h",
) 
``` 

Compiling with `bazel build //tensorflow/lite/examples/systemc:sc_example_test` produces the error:
```
ERROR: /working_dir/tensorflow/tensorflow/lite/examples/systemc/BUILD:63:1: C++ compilation of rule '//tensorflow/lite/examples/systemc:sc_example_test' failed (Exit 1)
In file included from tensorflow/lite/examples/systemc/sc_gtest_test.sc.cc:6:0:
bazel-out/k8-opt/bin/external/systemc/systemc-2.3.3/install/include/systemc/systemc.h:219:19: fatal error: systemc: No such file or directory
compilation terminated.
```

Limiting the number of jobs to 1: `bazel build --jobs 1 //tensorflow/lite/examples/systemc:sc_example_test` yields a successful compilation.


Contents of the folder where bazel copies to

```
/working_dir/tensorflow/bazel-bin/external/systemc
`-- systemc-2.3.3
    `-- install
        |-- include
        |   `-- systemc
        |       `-- systemc.h   <<< missing on bad compilation
        `-- lib
            `-- libsystemc.a  <<< missing on bad compilation
```

Contents of the folder copied by bazel (systemc cmake build/install folder)

```
/working_dir/tensorflow/bazel-tensorflow/external/systemc/systemc-2.3.3/install
|-- include
|   `-- systemc
|       |-- sysc
|       |-- systemc
|       |-- systemc.h
|       |-- tlm
|       |-- tlm.h
|       |-- tlm_core
|       `-- tlm_utils
|-- lib
|   |-- cmake
|   |   |-- SystemCLanguage
|   |   `-- SystemCTLM
|   `-- libsystemc.a
```

Somehow, a dependent task executes before the the genrule finishes when `--jobs 1` is not set.
Any insights?

**Comment by iirina** [comment]
Thanks for providing all the details. I'm forwarding this to the C++ team, maybe they have some ideas about what's going on.

**Comment by sgowroji** [comment]
Hi there! We're doing a clean up of old issues and will be closing this one. Please reopen if you’d like to discuss anything further. We’ll respond as soon as we have the bandwidth/resources to do so.
